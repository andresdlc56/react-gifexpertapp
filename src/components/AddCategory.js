import React, { useState } from 'react';

//Libreria necesaria para la validacion de los props
import PropTypes from 'prop-types';


/* ====================================

    ESTE ES EL SEGUNDO COMPONENTE (HIJO)
    MANEJA EL INPUT 

    ESTE COMPONENTE MODIFICA EL STATE DEL SU COMPONENTE PADRE

======================================= */


export const AddCategory = ({ setCategories }) => {
    
    const [inputValue, setValue] = useState('');


    // DEFINIENDO EVENTO QUE SE VA A DISPARAR CADA VEZ QUE SE TECLEA ALGO EN EL INPUT 
    const handleInputChange = ({ target }) => {

        //CAMBIANDO EL VALOR DEL useState "inputValue" POR EL VALOR DEL TARGET
        setValue(target.value);
    }

    
    // MANEJADOR DEL ENVIO DEL FORMULARIO
    const handleSubmit = (e) => {

        // EVITANDO EL REFRESCAMIENTO DE LA PANTALLA
        e.preventDefault();

        // VALIDACION 
        if(inputValue.trim().length > 2){
            /* DEBIDO A QUE NO TENGO ACCESO A LAS "categories" DIRECTAMENTE DESDE ESTE COMPONENTE POR QUE 
            NO LO PEDI EN LOS PROPS HAGO UNA REFERENCIA DEL MISMO */
            setCategories((cats) => [
                inputValue,
                ...cats
            ]);

            //LIMPIANDO EL VALOR DE INPUT
            setValue('');
        }

        
    }
    
    return (
        <form onSubmit={ handleSubmit }>
            <p>{ inputValue }</p>
            <input 
                type='text'
                value={ inputValue }
                onChange={ handleInputChange }
            />
        </form>  
    )
}


AddCategory.protoType = {
    setCategories: PropTypes.func.isRequired
}