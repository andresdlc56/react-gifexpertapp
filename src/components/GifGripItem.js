import React from 'react';

import PropTypes from 'prop-types';

export const GifGripItem = ({ title, url }) => {

    return (
        <div className="card animate__animated animate__bounce">
            <img src={ url } alt={ title }/>
            <p>{ title }</p>
        </div>
    )
}

GifGripItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}