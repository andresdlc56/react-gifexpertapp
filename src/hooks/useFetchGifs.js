
/*============================================== 

    LOS CUSTON HOOK, SON HOOKS DEFINIDOS POR EL PROGRAMADOR 
    PARA MANEJAR PROCESOS QUE SE REPETIRAN CONTINUAMENTE 

    PARATICULARMENTE ESTE HOOK SE ENCARGAR DE MANEJAR LAS PETICIONES
    A LA API DE GIFS

========================================== */

import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = (category) => {
    
    //ESTE SERA EL ESTADO POR DEFECTO DE ESTE HOOK
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    /* EFECTO QUE SE ACTIVARA CADA VEZ QUE LA CATEGORIA CAMBIE */
    useEffect(() => {

        getGifs(category)
            .then(gifs => {
                setTimeout(() => {
                    setState({
                        data: gifs,
                        loading: false
                    })
                }, 3000);
            })
        
        
       
    }, [category])
    
    

    return state;
}
