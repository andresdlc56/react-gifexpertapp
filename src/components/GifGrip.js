import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGripItem } from './GifGripItem';

//import { getGifs } from '../helpers/getGifs';

import PropTypes from 'prop-types';

export const GifGrip = ({ category }) => {

    /* -------------------------------
        REACT POR NATURALEZA VUELVE A LEER Y EJECUTAR TODO EL COMPONENTE
        AL NOTAR ALGUN CAMBIO EN SI 
    -------------------------------- */


    // IR A BUSCAR GIFS A LA API
    const { data, loading } = useFetchGifs(category);
    

    return (
        <>
            
        
            {
                loading
                    ? <p>LOADING...</p>
                    : 
                    <div>
                        <h3>{ category }</h3>
                        <div className="card-grid">
                            {
                                data.map((img) => (
                                    <GifGripItem 
                                        key={ img.id }
                                        { ...img }
                                    />
                                ))
                            } 
                        </div>
                    </div>
                    
            }

            
        </>
    )
}

GifGrip.protoType = {
    category: PropTypes.string.isRequired
}