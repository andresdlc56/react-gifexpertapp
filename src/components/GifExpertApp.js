import React, { useState } from 'react';
import { AddCategory } from './AddCategory';
import { GifGrip } from './GifGrip';


/* ====================================

    ESTE ES EL COMPONENTE PRINCIAL (PADRE)

======================================= */

export const GifExpertApp = ({ defaultCategories = [] }) => {


    // CREANDO UN useState PARA MANEJAR LAS CATEGORIAS 
    //const [categories, setCategories] = useState(['Dragon Ball']);
    const [categories, setCategories] = useState(defaultCategories);


    return (
        <>
            <h2>GifExpertApp</h2> 

            <AddCategory setCategories={ setCategories }/>

            <hr/> 


            <ol>
                {
                    // BARRIENDO EL ARREGLO DE CATEGORIAS
                    categories.map((category) => ( 
                        <GifGrip 
                            key={ category }
                            category={ category }
                        />
                    ))
                }
            </ol>

          
        </>
    )
}
