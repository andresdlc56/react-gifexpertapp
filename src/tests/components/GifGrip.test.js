import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from "enzyme";
import { GifGrip } from '../../components/GifGrip';
import { useFetchGifs } from '../../hooks/useFetchGifs';

jest.mock('../../hooks/useFetchGifs'); // FINGIR LLAMADOS A ESE ARCHIVO (IMPORTANTISIMO)

describe('Pruebas en el componente <GifGrip />', () => {

    test('Debe mostrar el componente correctamente', () => {

        // SIMULANDO QUE EL CUSTON HOOK useFetchGifs() YA CARGO LA INFO
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow(<GifGrip category='Dragon Ball' />);

        expect(wrapper).toMatchSnapshot();
    });


    test('Debe mostrar items cuando se cargan los gifs con el hook useFetchGifs()', () => {
        
        // SIMULACION DE DATA DEVUELTA POR LA API GIFS
        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/cualquier/cosa.jpg',
            title: 'Cualquier Cosa'
        },
        {
            id: '123',
            url: 'https://localhost/cualquier/cosa.jpg',
            title: 'Cualquier Cosa'
        }];

        // SIMULANDO QUE EL CUSTON HOOK useFetchGifs() YA CARGO LA INFO
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });
        
        const wrapper = shallow(<GifGrip category='Dragon Ball' />);


        
        
        // VERIFICANDO QUE ESTE PARRAFO NO EXISTA EN EL COMPONENTE
        expect(wrapper.find('p').exists()).toBe(false);
        
        /* VERIFICANDO QUE EL COMPONENTE GifGripItem EXISTE LA MISMA CANTIDAD
        DE VECES QUE GIFS ENCONTRADOS */
        expect(wrapper.find('GifGripItem').length).toBe(gifs.length);
        
    });
    
});