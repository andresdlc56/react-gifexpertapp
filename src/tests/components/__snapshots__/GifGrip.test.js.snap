// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <GifGrip /> Debe mostrar el componente correctamente 1`] = `
<Fragment>
  <p>
    LOADING...
  </p>
</Fragment>
`;
